src/era/rotating.py

```diff
@@ -0,0 +1,106 @@
+"""Create a new directory in base_dir and delete old directories as needed, logging as we go.
+"""
+
+import logging
+import re
+import shutil
+from datetime import datetime
+
+module_logger = logging.getLogger(__name__)
+
+
+class FolderRotator:  # pylint: disable=too-few-public-methods
+    """Create a new directory in base_dir and delete old directories as needed, logging as we go.
+    """
+
+    def __init__(self, base_dir) -> None:
+        self.base_dir = base_dir
+        self._class_logger = logging.getLogger(__name__).getChild(self.__class__.__name__)
+
+    #: Make dir Methods
+    def _build_new_download_dir_path(self, prefix, date_format):
+        today = datetime.today()
+        download_dir = self.base_dir / f'{prefix}{today.strftime(date_format)}'
+        return download_dir
+
+    def _make_new_download_dir(self, download_dir_path, exist_ok):
+        self._class_logger.debug(f'Attempting to create new directory `{download_dir_path}`')
+        try:
+            download_dir_path.mkdir(exist_ok=exist_ok)
+        except FileNotFoundError as error:
+            raise FileNotFoundError(f'Base directory `{self.base_dir}`` does not exist.') from error
+        else:
+            self._class_logger.debug(f'Successfully created `{download_dir_path}`')
+            return download_dir_path
+
+    #: Rotator Methods
+    def _get_all_but_n_most_recent_folder_paths(self, prefix, pattern, max_folder_count):
+        pattern = f'{prefix}{pattern}'
+        self._class_logger.debug(f'Folder regex pattern: {pattern}')
+        folder_paths = [path for path in self.base_dir.iterdir() if re.match(pattern, path.stem)]
+        if max_folder_count > len(folder_paths):
+            self._class_logger.debug(
+                f'max_folder_count `{max_folder_count}` greater than number of existing folders `{len(folder_paths)}`; '
+                'no folders deleted'
+            )
+            max_folder_count = 0
+        return sorted(folder_paths)[:-max_folder_count]
```

Comment:
**question**: `folder_paths` is a `Path` object in this place right. Does it sort based on string semantics?  Would you ever want to use os level folder metadata for knowing the age of a folder or could that be problematic?

Reply:
AFIU, it sorts based on lexicographic order, which means it takes into account the parent directories before sorting purely by alphabetical order. 

I've always shied away from trusting filesystem-level metadata for dates because I've no guarantee that some other process or user hasn't inadvertently changed the attribute I'm depending on. It's probably not that likely or hazardous here, but it's just kind of my default to manually encode that info in the filename. That way I/someone else would have to be very deliberate in changing it.

The broader implication of the `date_format` and regex `pattern` is intended to ensure any folder deletion has as narrow a scope as possible. If someone put another folder in the directory, it's an unlikely pattern for human-named folders so it won't grab it "by accident". Also, it allows someone to make a copy or rename a folder they want to keep around for testing or data archiving by changing the folder name.  

That's my thought process, anyways.